### PR TITLE
Add meta kernel for ctc_loss.intList

### PR DESCRIPTION
**Pull request by kundaMwiza**

### CODE REVIEW

**Comment by zou3519** [comment]
Is there a reason why you want to add a C++ meta instead of a Python meta?

**Comment by kundaMwiza** [comment]
@zou3519 The code was already there, I just refactored`ctc_loss_template_cpu` to move the shape inference to another function

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -1967,6 +1967,7 @@
   dispatch:
     CPU: ctc_loss_cpu
     CUDA: ctc_loss_gpu
+    Meta: ctc_loss_meta
```
Comment (zou3519) [inline aten/src/ATen/native/native_functions.yaml:1970]: Can you add a test for this?

File: aten/src/ATen/native/LossCTC.cpp [hunk 2]
```diff
@@ -116,6 +110,35 @@ std::tuple<Tensor, Tensor> ctc_loss_cpu_template(const Tensor& log_probs, const
...
+  constexpr scalar_t neginf = -std::numeric_limits<scalar_t>::infinity();
+  using target_t = typename std::conditional<target_scalar_type == kInt, int, int64_t>::type;
+
+  Tensor neg_log_likelihood, log_alpha;
+  size_t tg_target_stride;
+  std::vector<int64_t> tg_batch_offsets;
+
+  if (targets.scalar_type() == kLong) {
+    std::tie(neg_log_likelihood, log_alpha,tg_target_stride, tg_batch_offsets) =
+        ctc_loss_meta_template<scalar_t, kLong>(
```
Comment (zou3519) [inline aten/src/ATen/native/LossCTC.cpp:133]: ctc_loss_meta_template -> we should name this something better. Maybe `ctc_loss_allocate_outputs`?

**Review by zou3519 — COMMENTED** [review]
Code LGTM, could we add a test?

File: test/test_meta.py [hunk 3]
```diff
@@ -541,6 +541,10 @@ def run_meta_crossref(
                 else:
                     indices.append(meta_index)
             meta_args = (meta_args[0], indices)
+        elif func is torch.nn.functional.ctc_loss and all([isinstance(args[2], list), isinstance(args[3], list)]):
+            # torch.ops.aten._ctc_loss.IntList has a meta kernel but
```
Comment (zou3519) [inline test/test_meta.py:545]: the list check isn't really robust enough, someone could pass in a list of Tensors. Could we have a helper function check for list of ints?

**Review by zou3519 — COMMENTED** [review]
Minor nit about the testing. Otherwise, this looks good to me.

File: test/test_meta.py (hunk 3 above)
Comment (kundaMwiza) [reply test/test_meta.py:545]: This code will not be reached in that case since the only compatible signatures are those below?

```yaml
- func: ctc_loss.IntList(Tensor log_probs, Tensor targets, int[] input_lengths, int[] target_lengths, int blank=0, int reduction=Mean, bool zero_infinity=False) -> Tensor

# convenience function that converts to intlists for you
- func: ctc_loss.Tensor(Tensor log_probs, Tensor targets, Tensor input_lengths, Tensor target_lengths, int blank=0, int reduction=Mean, bool zero_infinity=False) -> Tensor
```

Comment (zou3519) [reply test/test_meta.py:545]: Fair enough, thanks for pointing that out

**Comment by zou3519** [comment]
btw, test failures look real

**Comment by kundaMwiza** [comment]
@zou3519 The previous test failures were due to providing int[] inputs for input_lengths and target_lengths and also setting reduction = "Mean". The mean is computed as below. However `get_clamped_target_length` creates an empty tensor, copies target_lengths values into that tensor via std::copy, and finally calling clamp on the resulting tensor. However the call to std::copy is not composite compliant since a data pointer is required from a tensor. This causes composite compliance tests to fail. Similarly some fx tests fail for the same reason since the std::copy call is not captured by the trace

```c++
    auto target_lengths_t = get_clamped_target_length(target_lengths, res.options());
    return (res / target_lengths_t).mean();
```

**Comment by zou3519** [comment]
To clarify:
- You're adding a meta for aten::_ctc_loss, which is not a Composite operator
- When running aten::ctc_loss.IntList (which is a Composite operator), it needs to construct a new Tensor from some IntList data
- constructing the new Tensor from the IntList data causes the tests to fail due to needing to write data into the new Tensor?

**Comment by kundaMwiza** [comment]
@zou3519 correct. The meta is added for aten._ctc_loss (really it is aten._ctc_loss.IntList). It is possible to add a meta kernel for this since int[] data can be read from for shape inference but not for aten._ctc_loss.Tensor. 

The recent failures are due to the addition of new sample inputs to the ctc loss OpInfo. These new sample inputs return input and target lengths of type int[] in order to actually test aten._ctc_loss meta. Previously the sample inputs only allowed aten._ctc_loss.Tensor to be called   Which doesn't have a meta kernel.

The side effect of adding these new sample inputs is that it caused the composite compliance tests for ctc loss to fail due to requiring converting int[] -> Tensor and requiring Tensor.data_ptr in order to write to it. I fixed this by not generating these problematic sample inputs but I suppose the alternative is to mark such tests as an xfail?

**Comment by zou3519** [comment]
Tbh, constructing the new Tensor from the IntList data should be allowed. At the callsite where we're doing it, we likely just need to exclude a number of DispatchKeys (https://github.com/pytorch/pytorch/blob/8dedc9dd9b8358899a740357e5375378d56b0bff/torch/csrc/utils/tensor_new.cpp#L368-L393). Doesn't need to happen in this PR.

> The side effect of adding these new sample inputs is that it caused the composite compliance tests for ctc loss to fail due to requiring converting int[] -> Tensor and requiring Tensor.data_ptr in order to write to it. I fixed this by not generating these problematic sample inputs but I suppose the alternative is to mark such tests as an xfail?

Not generating the sample inputs (assuming that was the state before this PR) is fine. Alternatively, you could have a separate OpInfo with the new sample inputs and then we would mark all such tests as xfail.

**Review by zou3519 — APPROVED** [review]

**Comment by kundaMwiza** [comment]
@zou3519 Thanks for approving. I'm unable to merge the PR as I'm not authorized.
